Example project/main.c

```diff
@@ -1,68 +1,137 @@
 #define TIM4_PERIOD       124
 
-void clk_config(void);
-void TIM4_Config(void);
-void delay()
-{
-  for(u16 i = 0; i < 0xFFFF; i++){}
-}
-void main( void )
-{  
-clk_config();
-TIM4_Config();
-GPIO_Init (GPIOB, GPIO_PIN_5, GPIO_MODE_OUT_PP_LOW_FAST);
-GPIO_Init (GPIOD, GPIO_PIN_3, GPIO_MODE_OUT_PP_LOW_FAST);
+/* Private macro -------------------------------------------------------------*/
 
-enableInterrupts();
+/* Private variables ---------------------------------------------------------*/
 
-while(1)
-{ 
-rainbowCycle(25); 
-}
-}
+/* Private function prototypes -----------------------------------------------*/
+static void CLK_Config(void);
+static void TIM4_Config(void);
+static void GPIO_Config(void);
+void _delay_ms(u16);
+void uptime_routine(void);
 
+/* Private user code ---------------------------------------------------------*/
+void main( void )
+{  
+  /* MCU Configuration -------------------------------------------------------*/
+  CLK_Config();
+  TIM4_Config();
+  GPIO_Config();
+  
+  enableInterrupts();
+  
+  /* Infinite Loop -----------------------------------------------------------*/
+  while(1)
+  { 
+    rainbowCycle(25); 
+  }
+}
 
-void clk_config(void)
+/* Clock Init ----------------------------------------------------------------*/
+/**
+  * @brief  Configure system clock to run at Maximum clock speed and output the 
+  *         system clock on CCO pin
+  * @param  None
+  * @retval None
+  */
+static void CLK_Config(void)
 {
 CLK_HSECmd(ENABLE);
 CLK_ClockSwitchConfig(CLK_SWITCHMODE_MANUAL, CLK_SOURCE_HSE, DISABLE, CLK_CURRENTCLOCKSTATE_DISABLE);
 CLK_ClockSwitchCmd(ENABLE);
 }
 
-void TIM4_Config(void)
+/* GPIO Init -----------------------------------------------------------------*/
+/**
+  * @brief  Configure GPIO Pins 
+  * @param  None
+  * @retval None
+  */
+static void GPIO_Config(void)
 {
-  TIM4_TimeBaseInit(TIM4_PRESCALER_128, TIM4_PERIOD);
-  TIM4_ClearFlag(TIM4_FLAG_UPDATE);
-  TIM4_ITConfig(TIM4_IT_UPDATE, ENABLE);
-  TIM4_Cmd(ENABLE);
+  GPIO_Init (GPIOB, GPIO_PIN_5, GPIO_MODE_OUT_PP_LOW_FAST);
+  GPIO_Init (GPIOD, GPIO_PIN_3, GPIO_MODE_OUT_PP_LOW_FAST);
 }
 
+/* Timer4 Init ---------------------------------------------------------------*/
+/**
+  * @brief  Configure TIM4 peripheral   
+  * @param  None
+  * @retval None
+  */
+static void TIM4_Config(void)
+{
+  /* TIM4 configuration:
+   - TIM4CLK is set to 16 MHz, the TIM4 Prescaler is equal to 128 so the TIM1 counter
+   clock used is 16 MHz / 128 = 125 000 Hz
+  - With 125 000 Hz we can generate time base:
+      max time base is 2.048 ms if TIM4_PERIOD = 255 --> (255 + 1) / 125000 = 2.048 ms
+      min time base is 0.016 ms if TIM4_PERIOD = 1   --> (  1 + 1) / 125000 = 0.016 ms
+  - In this example we need to generate a time base equal to 1 ms
+   so TIM4_PERIOD = (0.001 * 125000 - 1) = 124 */
+
+  /* Time base configuration */
+  TIM4_TimeBaseInit(TIM4_Prescaler_128, TIM4_PERIOD);
```

Comment:
TIM4_PRESCALER_128, not TIM4_Prescaler_128. The same for TIM4_FLAG_Update, TIM4_IT_Update

Reply:
> TIM4_PRESCALER_128, not TIM4_Prescaler_128. The same for TIM4_FLAG_Update, TIM4_IT_Update (sorry for being sourish - just spent almost a day and started losing my hair on this)

Do we use the same SPL? I noticed, that ST was using inconsistent naming between the STM8S and STM8L libraries. Could it be that they changed it now?

Reply:
https://github.com/j3qq4hch/STM8_WS2812B/blob/6ebab04c059536a1bbcd1ea5925976b20190eb23/STM8S_StdPeriph_Driver/inc/stm8s_tim4.h#L57